gadopt/gplates.py

```diff
@@ -0,0 +1,449 @@
+            [pygplates.PointOnSphere(x=coords[i, 0], y=coords[i, 1], z=coords[i, 2], normalise=True)
+             for i in range(np.shape(coords)[0])]
+        )
+
+        # Create a feature containing the multipoint feature, and defined as MeshNode type
+        meshnode_feature = pygplates.Feature(pygplates.FeatureType.create_from_qualified_string('gpml:MeshNode'))
+        meshnode_feature.set_geometry(multi_point)
+        meshnode_feature.set_name('Velocity Mesh Nodes from pygplates')
+
+        output_feature_collection = pygplates.FeatureCollection(meshnode_feature)
+
+        return output_feature_collection
+
+    def _interpolate_seeds_u(self, target_coords):
+        """
+        Interpolates seed velocities onto target coordinates.
+
+        This method generates a KD-tree of seed points with numerical values, then uses
+        this tree to find the nearest neighbors of the target coordinates. It interpolates
+        the velocities of these neighbors to the target points, applying a weighted average
+        based on distance. If any target point is exceptionally close to a seed point,
+        it directly assigns the seed's velocity to avoid division by zero errors.
+
+        Args:
+            target_coords (numpy.ndarray): Array of target coordinates for velocity interpolation.
+
+        Returns:
+            numpy.ndarray: Interpolated velocities at the target coordinates.
+        """
+        # calculate velocities here
+        seeds_u = self._calc_velocities(
+            velocity_domain_features=self.velocity_domain_features,
+            topology_features=self.topology_features,
+            rotation_model=self.rotation_model,
+            time=round(self.reconstruction_time, 0),
+            delta_time=self.delta_time)
+
+        seeds_u = np.array([i.to_xyz() for i in seeds_u]) *\
+            ((1e-2 * pyGplatesConnector.velocity_non_dim_factor) / (self.scaling_factor * pyGplatesConnector.yrs2sec))
+
+        # generate a KD-tree of the seeds points that have a numerical value
+        tree = cKDTree(data=self.seeds[seeds_u[:, 0] == seeds_u[:, 0], :], leafsize=16)
+
+        # find the neighboring points
+        dists, idx = tree.query(x=target_coords, k=self.nneighbours)
+
+        # Use np.where to avoid division by zero
+        # Replace 0 distances with np.finfo(float).eps to avoid division
+        # by zero while keeping original zeros intact for later use
+        safe_dists = np.where(dists == 0, np.finfo(float).eps, dists)
+
+        # Then, calculate the weighted average using safe_dists
+        res_u = np.einsum(
+            'i, ij->ij',
+            1/np.sum(1/safe_dists, axis=1),
+            np.einsum('ij, ijk ->ik', 1/safe_dists, seeds_u[seeds_u[:, 0] == seeds_u[:, 0]][idx])
+        )
+
+        # Now handle the case where points are too close to each other:
+        res_u[dists[:, 0] <= 1e-8, :] = seeds_u[seeds_u[:, 0] == seeds_u[:, 0]][idx[dists[:, 0] <= 1e-8, 0]]
```

Comment:
`1e-8` here is a bit of a magic constant too.

I think this would be clearer if you assigned `dists[:,0] <= 1e-8` to its own variable.

Reply:
So this is basically to say, if we are extremely close to any point, let's not interpolate and assign the value itself. It is just to avoid non-sense results when dividing by zero distance. But a valid point re it being a magic constant. I define it as a constant now.